legend-shared-pac4j/src/main/java/org/finos/legend/server/pac4j/hazelcaststore/HazelcastSessionStore.java

```diff
@@ -0,0 +1,157 @@
+                    "Failed to find Hazelcast config file in specified path: " + hazelcastConfigFilePath, e);
+        }
+    }
+
+    private SessionToken getOrCreateSsoKey(WebContext context)
+    {
+        SessionToken token = SessionToken.fromContext(context);
+        if (token == null)
+        {
+            token = createSsoKey(context);
+        }
+        return token;
+    }
+
+    private SessionToken createSsoKey(WebContext context)
+    {
+        SessionToken token = SessionToken.generate();
+        token.saveInContext(context, maxSessionLength);
+        Map<String, Object> hazelcastSessionData = new HashMap<>();
+        hazelcastMap.put(token.getSessionId(), hazelcastSessionData);
+        return token;
+    }
+
+    @Override
+    public String getOrCreateSessionId(WebContext context)
+    {
+        getOrCreateSsoKey(context);
+        return super.getOrCreateSessionId(context);
+    }
+
+    @Override
+    public Object get(WebContext context, String key)
+    {
+        Object res = super.get(context, key);
+        if (res == null)
+        {
+            SessionToken token = getOrCreateSsoKey(context);
+            Map<String, Object> hazelcastSessionData = hazelcastMap.get(token.getSessionId());
+            if (hazelcastSessionData != null)
+            {
+                Object data = hazelcastSessionData.get(key);
+                if (data != null)
+                {
+                    res = data;
+                    //Once we have it, store it in the regular session store for later access
+                    super.set(context, key, res);
+                }
+            }
+        }
+        else if (SessionToken.fromContext(context) == null)
+        {
+            //if res is not null ,this means we still have an active Session but an expired SSO cookie we need to recreate one and add it to the context request/response.
+            createSsoKey(context);
+            if (res instanceof LinkedHashMap)
+            {
+                ProfileHelper.flatIntoAProfileList((LinkedHashMap<String, CommonProfile>)res);
+                set(context, Pac4jConstants.USER_PROFILES, ProfileHelper.flatIntoAProfileList((LinkedHashMap<String, CommonProfile>)res));
+            }
+        }
+        return res;
+    }
+
+    @Override
+    public void set(WebContext context, String key, Object value)
+    {
+        SessionToken token = getOrCreateSsoKey(context);
+        Map<String, Object> hazelcastSessionData = hazelcastMap.get(token.getSessionId());
+        if (hazelcastSessionData != null)
+        {
+            hazelcastSessionData.put(key, value);
+
+            // need to write the object back into hazelcast
+            hazelcastMap.put(token.getSessionId(), hazelcastSessionData);
+        }
+
+        super.set(context, key, value);
+    }
+
+    @Override
+    public boolean destroySession(WebContext context)
+    {
+        SessionToken token = getOrCreateSsoKey(context);
```

Comment:
It seems odd to create a session token here if one doesn't already exist. Not wrong, necessarily, but odd, since the purpose here is to destroy the session.